### PR TITLE
Adds seasonal variation to daylight levels

### PR DESCRIPTION
**Pull request by Toothspit**
Sunlight levels for New England lattitudes vary, with the summer solstice 23% more intense than equinox levels, and the winter solstice 30% less intense. I'm actively looking for a better source for the deviation, but I used the tools below to calculate it.

Currently this PR linearly adjusts sunlight level in calendar::sunlight() according to the day of the year, a good approximation for the ~sinusoidal change in elevation angle of the Sun.

We currently assume solstice and equinox timing from sunset() and sunrise() times, though it has been pointed out that temperature's max and min occur mid-season -- it's unsettled whether this is intentional or a "feature" to mimic the delay between i.e. solar irradiance maximum (summer solstice, ~22June) and surface temperature maximum (late-July).

To check my math (please do), see [the Wiki article](https://en.wikipedia.org/wiki/Effect_of_sun_angle_on_climate) for an explanation of the effect, [insolation fraction](http://www.debarel.com/Tools/insolation_calculator.html) is the ratio of sunlight on the ground compared to 'straight up,' and Wikipedia also has [solstice and equinox dates](https://en.wikipedia.org/wiki/Solstice) to use in [NOAA's calculator](https://www.esrl.noaa.gov/gmd/grad/solcalc/) for sun elevation.
*****
This can get kinda confusing, so I made graphs to help. [Here](http://imgur.com/aZH4fht.png)'s Boston, MA's real sunrise and sunset times at solstice and equinox for two years. [Here'](http://imgur.com/bGCA09n.png)s how DDA currently approximates those times.

Boston's real 'light level' at noon is [here](http://imgur.com/jWHW5G8.png), as a percentage of what it would be if the sun were straight up at the zenith. [Here](http://imgur.com/tNOUkq3.png)'s this PR's approximation of that effect.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
3% is very little, especially when it's not +3% but `*103%`.
Adding flat 3% of DAYLIGHT_LEVEL in summer and substracting it in winter would give more realistic effects in that it would also affect the nights.

**Comment by Toothspit** [comment]
I don't follow, where'd I screw up? I thought that's what I did, with `1.03 * DAYLIGHT_LEVEL` during summer? Should it be more like `DAYLIGHT_LEVEL + 3` instead? I made it multiplicative so changes to DAYLIGHT_LEVEL and SEASONAL_OFFSET would work without adjusting the other one (i.e. a DAYLIGHT_LEVEL of 400 but still x% variation instead of +3), but again, IANA real programmer.

DAYLIGHT_LEVEL shouldn't be affecting nights at all, L291 returns moonlight using MOONLIGHT_PER_QUARTER and current moon phase. Am I reading it wrong?

**Comment by Coolthulhu** [comment]
> I thought that's what I did, with 1.03 * DAYLIGHT_LEVEL during summer? Should it be more like DAYLIGHT_LEVEL + 3 instead?

I mean, just multiplying by 103% is very little effect. You'd barely notice it in the game.

**Comment by Coolthulhu** [comment]
Though I guess it would be somewhat harder to implement, since there is that percentage multiplication later.

A technical suggestion: wrap the multiplier in a function. All the other elements of the function (seconds after sunset etc.) are functions, which makes it more readable.

**Comment by Toothspit** [comment]
Yeah, I definitely get that this could be too minor an effect to warrant implementing/maintaining, but I figured I'd pop up a PR and see what folks thought. Should I add [CR]?

**Comment by Coolthulhu** [comment]
You could find data that exaggerates the effect somewhat or uses different units.
Realism is not objective - it's a matter of perception. And it and often requires some adjustments to work within the limitations of the system, because we only simulate some of the effects that affect sunlight IRL.

For example, if the data you used accounted for logarithmic scale of light intensity perception, it may work for full sunlight, but will not represent low sunlight levels too well. Or solar panel output.

File: src/calendar.cpp [hunk 1]
```diff
@@ -257,11 +257,33 @@ bool calendar::is_night() const
...
 }
 
+double calendar::current_daylight_level() const
+{
+    double current_daylight_modifier = 1.;
+    
+    switch (season) {
+    case SPRING:
+        current_daylight_modifier = 1.;
+        break;
```
Comment (codemime) [inline src/calendar.cpp]: Way too verbose. The whole function could be shortened:
```c++
double calendar::daylight_modifier()
{
    switch (season) {
    case SUMMER:
        return 1.03;
    case WINTER:
        return 0.97;
    }
    return 1.0;
}

...

double daylight_level = double(DAYLIGHT_LEVEL) * daylight_modifier();

```
Also, smooth changes would be better than sudden ones.

**Comment by Toothspit** [comment]
Specifically solar panels are a different thing, because different frequencies of sunlight get attenuated differently as they pass through the atmosphere at different angles, and panel efficiency is strongly influenced by received frequency. That's a super-hard problem, but we can abstract it by just adjusting SEASONAL_DAYLIGHT_OFFSET and/or DAYLIGHT_LEVEL and changing panels' epower (see vehicle.cpp L5953).

Currently solar panel efficiency/sunlight frequency components and all that is just abstracted away into epower output for total weather-adjusted sunlight level, so I assume W/m^2 is as good a metric as any since it's just a big glob of "how much solar radiation hits the ground per unit time."

I pulled that modifier out to the header specifically to make those abstractions easier to adjust for. Using sunlight() differently for eyeballs/cameras/panels is probably a different PR, though I agree with you they are *definitely* affected differently by different zenith angles.

**Comment by Coolthulhu** [comment]
> I pulled that modifier out to the header specifically to make those abstractions easier to adjust for.

Having modifiers in headers actually makes adjustments harder if they're only used in one place. That is because you need to recompile everything that included the header.

File: src/calendar.cpp (hunk 1 above)
Comment (Toothspit) [inline src/calendar.cpp]: I wasn't sure whether to be as explicit, I was just going off the style of the earlier functions, will use yours.

That sunlight() function is called every turn for vehicles with panels from what I can tell, do we really want to interpolate for a 3% change? I don't by any means know what I'm doing, but that sounds performance-heavy for cases where SEASONAL_DAYLIGHT_OFFSET is small.

**Comment by Toothspit** [comment]
I'm not really sure how to handle it then, hardcode the 3% variation into current_daylight_level()?

**Comment by Coolthulhu** [comment]
Hardcode is good for now. It can be considered an implementation detail.

Later on (not this PR) it should be moved to json to become adjustable.

**Comment by Toothspit** [comment]
I'm not sure if anyone can find a better source, but I just compared noon solar declination at the summer/winter solstices for Boston (71deg and 24deg from horizon), comparing the cosines to show how 'spread out' the light would be for a given piece of ground, and got a 58% difference in received light per area. The source I used above could not possibly have been for what we're trying to model.

Can we just pick a number? Math says we should be using 1.29 for summer and 0.71 for winter, but that would make seasonal transitions very sudden if we don't interpolate, which sounds performance-heavy for something that currently only affects solar panel output from what I can tell.

**Comment by Coolthulhu** [comment]
> Can we just pick a number?

It's a bit more complicated, but picking a number with a reasoning behind it is a good start.
> Math says we should be using 1.29 for summer and 0.71 for winter, but that would make seasonal transitions very sudden if we don't interpolate, which sounds performance-heavy for something that currently only affects solar panel output from what I can tell.

Linear interpolation isn't expensive unless you do a lot of it. And sunlight isn't calculated often - the only time it is is when calculating solar panel output retroactively. And when that happens, the weather also needs to be calculated and weather is much more complex (see weather_gen.cpp).

**Comment by Toothspit** [comment]
Looks kludgy, but with all the sign-flipping I settled for switching on season, couldn't figure out a cleaner way to handle it and keep the option to jsonize a single value later.

Still open to a better source than my own math for the variance.

File: src/calendar.cpp (hunk 1 above)
Comment (Toothspit) [inline src/calendar.cpp]: @Coolthulhu thinks the performance impact is small, so I implemented the smooth changes you suggested.

**Comment by Coolthulhu** [comment]
Is the light intensity supposed to be the highest in the middle of a season or at start?

Currently the temperature is the highest/lowest in the middle of summer/winter, so I'd expect light intensity to also have the extreme value in the middle.

**Comment by Toothspit** [comment]
My understanding from sunrise/set times as currently implemented is that the first day of a season is the solstice/equinox for that season, so all of spring has light intensity increasing from the spring equinox (middle point of intensity) to the summer solstice (highest point) then decreasing to the autumnal equinox (middle point) through to the winter equinox (lowest point) then back up to the spring equinox to start the year again.

Daylight hours are proportional to sunlight intensity, and sunrise() and sunset() seem to assume that the first day of spring is the spring equinox. I'm honestly not sure how temperatures work WRT solstice/equinox timing, but it *seems* like the hottest parts of summer are after June and the coldest parts of winter are after December, so the temperature behavior could be either working as intended or a "feature."

Unfortunately, SUNRISE_SOLSTICE and SUNSET_SOLSTICE are also incorrectly named and should be _EQUINOX based on their usage -- equinox sunrise and sunset times happen twice a year, solstice times only once. But that's a cosmetic thing.

[Here's](http://i.imgur.com/rVnpLtD.png) how sunrise/set times currently affect hours of daylight, which is proportional to changes in the sun's elevation, which is what determines incoming light intensity. I can assume here that the sunrise/set times are correct, or I can shift the model in this PR by half a season to follow temperature if we think the model is incorrect.

**Comment by Toothspit** [comment]
Need some implementation advice. ~BrainDead in IRC made me re-check my data, I was off a little on the math. Using Boston as our reference location, [this insolation calculator](http://www.debarel.com/Tools/insolation_calculator.html), [NOAA's solar noon data](https://www.esrl.noaa.gov/gmd/grad/solcalc/sunrise.html), and [Wiki's chart for 2017's solstices and equinoxes](https://en.wikipedia.org/wiki/Solstice).

Summer solstice should be 23% brighter, winter solstice should be 30% less bright. This means the rate of change in brightness is greater from autumn->spring than spring->autumn. I'm going to ignore that and keep using a single now-26.5% offset in both directions to make this less complicated and have a single variable to jsonize later, since it's a very minor feature and doesn't warrant complicating for pure realism.

Modelling the more extreme winter would just be a matter of adding and using another `deviation`, though, if desired.

**Comment by Coolthulhu** [comment]
Round it to a multiple of 5%. No need to go for extreme precision when we don't have extreme precision from the sources.

File: src/calendar.cpp [hunk 2]
```diff
@@ -257,11 +257,37 @@ bool calendar::is_night() const
     return (seconds > sunset_seconds + TWILIGHT_SECONDS || seconds < sunrise_seconds);
 }
 
+double calendar::current_daylight_level() const
+{
+    double percent = double(double(day) / season_length());
```
Comment (Coolthulhu) [inline src/calendar.cpp]: No need to cast it twice. That outer cast is redundant.

File: src/calendar.cpp [hunk 3]
```diff
@@ -257,11 +257,37 @@ bool calendar::is_night() const
...
+    double modifier = 1.0;
+    // For ~Boston: solstices are +/- 25% sunlight intensity from equinoxes
+    static double deviation = 0.25;
```
Comment (Coolthulhu) [inline src/calendar.cpp]: Should also be const. Or `constexpr`.

**Pull request merged**